### DISCUSSION TITLE
retrieve the parameter model of an API and use that to define the parameter model

### DISCUSSION THREAD

**Discussion by anirban-ray-publicis-sapient** [fastapi/fastapi#9524] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

###### Low Level Service

```python3
# low-level API

from fastapi import FastAPI
from pydantic import BaseModel


class LowLevelParameterModel(BaseModel):
    item_name: str


app = FastAPI()


@app.post("/items/")
async def create_item(item: LowLevelParameterModel):
    return item

```

###### High Level Service

```python3
# high level API - assume no access to external / low level API

from fastapi import FastAPI
from pydantic import BaseModel


class HighLevelParameterModel(BaseModel):
    store_name: str
    store_item: LowLevelParameterModel  # do not have access to ``LowLevelParameterModel`` in this codebase


app = FastAPI()


@app.post("/store_items/")
async def create_store_items(store_item: HighLevelParameterModel):
    return store_item
```

### Description

Consider two services: one is low level service, let's call it LOW, and the other one is high level, let's call it HIGH. The setup is the case that we are working on HIGH, and we'll try to create some wrappers over LOW. And we shall have no access at all to LOW's codebase from HIGH's codebase, all we can possibly have are the endpoints and the OpenAPI specification.

LOW has some very generic endpoints. The objective of HIGH is to take some user inputs, few on LOW inputs and few on some extra functionality in HIGH. Then HIGH will create a payload suitable for LOW by adding default values, and make the call to LOW. After LOW is done, return the response to use, possibly with some modification.

I want to know how to define the `BaseModel` of HIGH by nesting the `BaseModel` of LOW, to which I do not have direct access. While I can get the format from Swagger UI, I do not want to redefine the classes once again in HIGH. I want to reuse the definition already specified in LOW and use that in HIGH. How can I achieve the same?

The shared code tries to show a simple case of this. Here, LOW is about just items, and HIGH is about store and it's details, one of which is items. I want to know how to define `HighLevelParameterModel` using `LowLevelParameterModel`. Of course my actual use case won't be this simple, and `LowLevelParameterModel` will be nested at multiple levels with lots of custom `TypedDict`.

### Operating System

Linux

### FastAPI Version

0.95.1

### Python Version

3.10.11

### Additional Context

If it is of importance, LOW endpoints are behind `OAuth2`, in similar format as documented in tutorials.

**Comment by jgould22** [comment]
It sounds like you need to extract those types and models into a library and use the library in each of your projects.

Another question is how much work is HIGH actually doing? If it is just easing the crafting of requests for the LOW service it feels like HIGH would be better suited to be a client library which lives with its service and has access to its types.

**Reply by anirban-ray-publicis-sapient** [reply]
> Another question is how much work is HIGH actually doing?

It's doing plenty of work. There are actually LOW_1, LOW_2, etc. (these are basically different types of ML functionalities), and HIGH is the providing backend for a UI combining solutions as appropriate.

> It sounds like you need to extract those types and models into a library and use the library in each of your projects.

I like this idea. This should solve it, but given the size of the codebases, it'd be solid amount of work. So, if possible, a quick short-term workaround will be very welcome.

I was sort of hoping to access the type definitions "somehow" using `/openapi.json`. Since those are available in Swagger as well, I hoped it would be extractable some way, and then can be converted back to `BaseModel`.

**Reply by jgould22** [reply]
I do not know of a trivial way to convert from yaml / json to pydantic types.

Another option may be to add the LOW repo using git submodules. They can be painful to work with however.
